# MetaCAT

Metagenome Clustering and Association Tool.

## Examples

Suppose we have **two** cohorts: `cohort1` (discovery cohort) and `cohort2` (replication cohort).

`cohort1` has **two** individuals: `c1i1` and `c1i2`.

`cohort2` has **two** individuals: `c2i1` and `c2i2`.

A simulated metagenome `assembly.gz` was assemblied from `cohort1`.

## 01. Preparations

All reads were trimmed, and host DNA was removed by yourself.

Assemblies were constructed either for each individual separately or for all individuals together.

**Assume the following dependencies are in your system path!**

**Activate the environments first if the following programs are installed!**

- `bowtie2`
- `checkm2`
- `gtdbtk`
- `MetaCAT`
- `samtools`

**Number of threads is set to 8.**

```bash
export THREADS=8
```

## 02. Align Reads To Assembly

### Input

`assembly.gz`

`c1i1.1.gz` `c1i1.2.gz` `c1i2.1.gz` `c1i2.2.gz`

### Output

`c1i1.bam` `c1i2.bam`

```bash
bowtie2-build --threads ${THREADS} assembly.gz assembly
for i in c1i1 c1i2
do
bowtie2 -f -p ${THREADS} -x assembly -1 ${i}.1.fasta.gz -2 ${i}.2.fasta.gz -S ${i}.sam
samtools sort -l 9 --threads ${THREADS} -o ${i}.bam ${i}.sam
rm ${i}.sam
done
rm assembly.*.bt2*
```

## 03. Create Bam Index

Set the number of threads with `-t|--threads` to improve computation speed.

This option is only applicable to multiple BAM files.

If bam files are stored on an HDD instead of an NVMe drive,

it is recommended to reduce the number of threads (e.g., -t 20).

### Input

`c1i1.bam` `c1i2.bam`

### Output

`c1i1.bam.index` `c1i2.bam.index`

```bash
MetaCAT indexBam -t ${THREADS} --bam *.bam
```

## 04. Cluster Metagenomic Sequences Using MetaCAT

### Input

`assembly.gz`

`c1i1.bam` `c1i1.bam.index` `c1i2.bam` `c1i2.bam.index`

### Output

`c1i1.metacat.mapping` `c1i2.metacat.mapping`

`c1i1.metacat.*.fasta` `c1i2.metacat.*.fasta`

```bash
MetaCAT seed --fasta assembly.gz --output metacat.seed
for i in c1i1 c1i2
do
MetaCAT coverage --bam ${i}.bam --output ${i}.metacat.coverage
MetaCAT cluster --fasta assembly.gz --seed metacat.seed \
--coverage ${i}.metacat.coverage --output ${i}.metacat
done
rm *.bam *.index *.metacat.coverage metacat.seed
```

## 05. Estimate Quality

Suppose we have fasta formatted cluster files from multiple individuals / datasets, and generated by different methods.

Cluster files should be named in the following format: `Dataset.Program.ClusterID.fasta`

**Dataset** and **ClusterID** fields must contain no dots.

### Input

`*.metacat.*.fasta`

### Output

`metacat.checkm2.tsv`

`metacat.benchmarkRW.tsv`

`metacat.benchmarkRW.*.pdf`

```bash
MetaCAT checkm2 --fasta *.metacat.*.fasta --output metacat.checkm2.tsv
MetaCAT benchmarkRW --combine --checkm2 metacat.checkm2.tsv --output metacat.benchmarkRW
```

## 06. Cluster Annotation

`MetaCAT gtdbtk` will generate a `GTDB-Tk.msh` file during the first run.

This file is dataset-independent and only depends on the GTDB you use.

Therefore, it can be specified in subsequent runs with `--mash-db GTDB-Tk.msh` to skip this step.

### Input

`*.metacat.*.fasta`

`metacat.checkm2.tsv`

### Output

`metacat.gtdbtk.tsv`

`GTDB-Tk.msh`

Annotate HQ clusters (contamination <= 0.10 and completeness >= 0.70):

```bash
MetaCAT gtdbtk --checkm2 metacat.checkm2.tsv --fasta *.metacat.*.fasta --output metacat.gtdbtk.tsv
```

## 07. Select Representatives

Representatives should be selected from all identified clusters.

### Input

`metacat.checkm2.tsv`

`metacat.gtdbtk.tsv`

`*.metacat.*.fasta`

### Output

`metacat.representative.annotation`

`metacat.representative.assembly`

`metacat.representative.mapping`

```bash
MetaCAT representative --fasta *.metacat.*.fasta --checkm2 metacat.checkm2.tsv \
--gtdbtk metacat.gtdbtk.tsv --output metacat.representative
```

## 08. Align Reads To Representative Metagenome

All reads should be aligned to the representative metagenome.

### Input

`c1i1.1.gz` `c1i1.2.gz` `c1i2.1.gz` `c1i2.2.gz`

`c2i1.1.gz` `c2i1.2.gz` `c2i2.1.gz` `c2i2.2.gz`

### Output

`c1i1.bam` `c1i2.bam`

`c2i1.bam` `c2i2.bam`

```bash
bowtie2-build --threads ${THREADS} metacat.representative.assembly metacat.representative
for i in c1i1 c1i2 c2i1 c2i2
do
bowtie2 -f -p ${THREADS} -x metacat.representative -1 ${i}.1.fasta.gz -2 ${i}.2.fasta.gz -S ${i}.sam
samtools sort -l 9 --threads ${THREADS} -o ${i}.bam ${i}.sam
rm ${i}.sam
done
rm metacat.representative.*.bt2*
```

## 09. Create Bam Index

### Input

`c1i1.bam` `c1i2.bam`

`c2i1.bam` `c2i2.bam`

### Output

`c1i1.bam.index` `c1i2.bam.index`

`c2i1.bam.index` `c2i2.bam.index`

```bash
MetaCAT indexBam --threads ${THREADS} --bam *.bam
```

## 10. Estimate Relative Abundance

### Input

`c1i1.bam` `c1i1.bam.index`

`c1i2.bam` `c1i2.bam.index`

### Output

`metacat.abundance-c1.tsv` `metacat.abundance-c2.tsv`

```bash
MetaCAT abundance --annotation metacat.representative.annotation --bam c1i1.bam c1i2.bam \
--mapping metacat.representative.mapping --output metacat.abundance-c1.tsv
MetaCAT abundance --annotation metacat.representative.annotation --bam c2i1.bam c2i2.bam \
--mapping metacat.representative.mapping --output metacat.abundance-c2.tsv
```

## 11. Abundance Test

### Input

`metacat.abundance-c1.tsv`

`Group.tsv`

### Output

`metacat.abundanceTest-c1.*.pdf`

```bash
MetaCAT abundanceTest --abundance metacat.abundance-c1.tsv --group Group.tsv --output metacat.abundanceTest-c1
```

## 12. Call Variants

### Input

`metacat.representative.annotation` `metacat.representative.mapping`

`c1i1.bam` `c1i1.bam.index` `c1i2.bam` `c1i2.bam.index`

`c2i1.bam` `c2i1.bam.index` `c2i2.bam` `c2i2.bam.index`

### Output

`metacat.variant-c1.tsv` `metacat.variant-c2.tsv`

```bash
MetaCAT variant --annotation metacat.representative.annotation --bam c1i1.bam c1i2.bam \
--mapping metacat.representative.mapping --output metacat.variant-c1.tsv
MetaCAT variant --annotation metacat.representative.annotation --bam c2i1.bam c2i2.bam \
--mapping metacat.representative.mapping --output metacat.variant-c2.tsv
rm *.bam.index
```

## 13. Metagenome-wide Association Analysis

### Input

`Phenotype.tsv`

`Covariate.tsv`

`metacat.abundance-c1.tsv`

`metacat.variant-c1.tsv`

### Output

`metacat.mwas-c1.tsv`

```bash
MetaCAT mwas --abundance metacat.abundance-c1.tsv --phenotype Phenotype.tsv \
--covariate Covariate.tsv --variant metacat.variant-c1.tsv --output metacat.mwas-c1.tsv
```

## 14. Metagenome-wide Association Analysis Plots

### Input

`metacat.mwas-c1.tsv`

### Output

`metacat.mwas-c1.qq.pdf` `metacat.mwas-c1.manhattan.pdf`

```bash
MetaCAT plotMWAS --mwas metacat.mwas-c1.tsv -o metacat.mwas-c1
```

## 15. Replicaton Analysis In An Independent Cohort

Only SNPs that are significant in the discovery cohort are tested in the replication cohort.

### Input

`Phenotype.tsv`

`Covariate.tsv`

`metacat.abundance-c2.tsv`

`metacat.variant-c2.tsv`

`metacat.mwas-c1.tsv`

### Output

`metacat.mwas-c2.tsv`

```bash
MetaCAT mwas --abundance metacat.abundance-c2.tsv --phenotype Phenotype.tsv \
--covariate Covariate.tsv --variant metacat.variant-c2.tsv --output metacat.mwas-c2.tsv --mwas metacat.mwas-c1.tsv
```
