# MetaCAT

Metagenome Clustering and Association Tool.

## Examples

Suppose we have **two** cohorts: `cohort1` (discovery cohort) and `cohort2` (replication cohort).

`cohort1` has **two** individuals: `c1i1` and `c1i2`.

`cohort2` has **two** individuals: `c2i1` and `c2i2`.

A simulated metagenome `metagenome.gz` was assemblied from `cohort1`.

## 01. Preparations

All reads were trimmed, and host DNA was removed by yourself.

Assemblies were constructed either for each individual separately or for all individuals together.

**Assume the following dependencies are on your system path.**

- `bowtie2`
- `checkm2`
- `gtdbtk`
- `metacat`
- `samtools`

**Number of threads is set to 8.**

```bash
export THREADS=8
```

## 02. Align Reads To Assembly

### Input

`metagenome.gz`

`c1i1.1.gz` `c1i1.2.gz` `c1i2.1.gz` `c1i2.2.gz`

### Output

`c1i1.bam` `c1i2.bam`

```bash
bowtie2-build --threads ${THREADS} metagenome.gz metagenome.gz
for i in c1i1 c1i2
do
bowtie2 -f -p ${THREADS} -x metagenome.gz -1 ${i}.1.fasta.gz -2 ${i}.2.fasta.gz -S ${i}.sam
samtools sort -l 9 --threads ${THREADS} -o ${i}.bam ${i}.sam
rm ${i}.sam
done
rm metagenome.gz.*.bt2*
```

## 03. Create Bam Index

Set the number of threads with `-t|--threads` to improve computation speed.

This option is only applicable to multiple BAM files.

### Input

`c1i1.bam` `c1i2.bam`

### Output

`c1i1.bam.index` `c1i2.bam.index`

```bash
MetaCAT indexBam --input *.bam
```

## 04. Cluster Metagenomic Sequences Using MetaCAT

### Input

`metagenome.gz`

`c1i1.bam` `c1i1.bam.index` `c1i2.bam` `c1i2.bam.index`

### Output

`c1i1.metacat.mapping` `c1i2.metacat.mapping`

`c1i1.metacat.*.fasta` `c1i2.metacat.*.fasta`

```bash
MetaCAT seed \
--fasta metagenome.gz \
--output metagenome.metacat.seed
for i in c1i1 c1i2
do
MetaCAT coverage \
--bam ${i}.bam \
--output ${i}.metacat.coverage
MetaCAT cluster \
--fasta metagenome.gz \
--seed metagenome.metacat.seed \
--coverage ${i}.metacat.coverage \
--output ${i}.metacat
done
rm *.bam *.index *coverage metagenome.metacat.seed
```

## 05. Estimate Quality

Suppose we have fasta formatted cluster files from multiple individuals / datasets, and generated by different methods.

Cluster files should be named in the following format: `Dataset.Program.ClusterID.fasta`

**Dataset** and **ClusterID** fields must contain no dots.

### Input

`*.metacat.*.fasta`

### Output

`CheckM2.tsv`

`BenchmarkRW.tsv`

`BenchmarkRW.*.pdf`

```bash
MetaCAT checkm2 \
--input *.metacat.*.fasta \
--output CheckM2.tsv
MetaCAT benchmarkRW \
--combine \
--input CheckM2.tsv \
--output BenchmarkRW
```

## 06. Cluster Annotation

`MetaCAT gtdbtk` will produce `GTDB-Tk.msh` file on the first run, which can be specified on subsequent runs with `--mash-db`.

### Input

`*.metacat.*.fasta`

`CheckM2.tsv`

### Output

`GTDB-Tk.tsv`

`GTDB-Tk.msh`

Annotate HQ clusters (contamination <= 0.10 and completeness >= 0.70):

```bash
MetaCAT gtdbtk \
--checkm2 CheckM2.tsv \
--input *.metacat.*.fasta \
--output GTDB-Tk.tsv
```

## 07. Select Representatives

Representatives should be selected from all identified clusters.

### Input

`CheckM2.tsv`

`GTDB-Tk.tsv`

`*.metacat.*.fasta`

### Output

`Representatives.annotation`

`Representatives.assembly`

`Representatives.mapping`

```bash
MetaCAT representative \
--fasta *.metacat.*.fasta \
--checkm2 CheckM2.tsv \
--gtdbtk GTDB-Tk.tsv \
--output Representatives
```

## 08. Align Reads To Representative Metagenome

All reads should be aligned to the representative metagenome.

### Input

`c1i1.1.gz` `c1i1.2.gz` `c1i2.1.gz` `c1i2.2.gz`

`c2i1.1.gz` `c2i1.2.gz` `c2i2.1.gz` `c2i2.2.gz`

### Output

`c1i1.bam` `c1i2.bam`

`c2i1.bam` `c2i2.bam`

```bash
bowtie2-build --threads ${THREADS} Representatives.assembly Representatives.assembly
for i in c1i1 c1i2 c2i1 c2i2
do
bowtie2 -f -p ${THREADS} -x Representatives.assembly -1 ${i}.1.fasta.gz -2 ${i}.2.fasta.gz -S ${i}.sam
samtools sort -l 9 --threads ${THREADS} -o ${i}.bam ${i}.sam
rm ${i}.sam
done
rm Representatives.assembly.*.bt2*
```

## 09. Create Bam Index

### Input

`c1i1.bam` `c1i2.bam`

`c2i1.bam` `c2i2.bam`

### Output

`c1i1.bam.index` `c1i2.bam.index`

`c2i1.bam.index` `c2i2.bam.index`

```bash
MetaCAT indexBam \
--input *.bam
```

## 10. Estimate Relative Abundance

### Input

`c1i1.bam` `c1i1.bam.index`

`c1i2.bam` `c1i2.bam.index`

### Output

`Abundance.tsv`

```bash
MetaCAT abundance \
--annotation Representatives.annotation \
--bam c1i1.bam c1i2.bam \
--mapping Representatives.mapping \
--output C1-Abundance.tsv
MetaCAT abundance \
--annotation Representatives.annotation \
--bam c2i1.bam c2i2.bam \
--mapping Representatives.mapping \
--output C2-Abundance.tsv
```

## 11. Abundance Test

### Input

`C1-Abundance.tsv`

`Group.tsv`

### Output

`C1-Test.*.pdf`

```bash
MetaCAT abundanceTest \
--abundance C1-Abundance.tsv \
--group Group.tsv \
--output C1-Test
```

## 12. Call Variants

### Input

`c1i1.bam` `c1i1.bam.index` `c1i2.bam` `c1i2.bam.index`

`c2i1.bam` `c2i1.bam.index` `c2i2.bam` `c2i2.bam.index`

### Output

`C1-Variant.tsv`

`C2-Variant.tsv`

```bash
MetaCAT variant \
--input c1i1.bam c1i2.bam \
--output C1-Variant.tsv
MetaCAT variant \
--input c2i1.bam c2i2.bam \
--output C2-Variant.tsv
rm *.bam.index
```

## 13. Metagenome-wide Association Analysis

### Input

`Representatives.mapping`

`Representatives.annotation`

`C1-Abundance.tsv`

`Phenotype.tsv`

`Covariate.tsv`

`C1-Variant.tsv`

### Output

`C1-MWAS.tsv`

```bash
MetaCAT mwas \
--abundance C1-Abundance.tsv \
--phenotype Phenotype.tsv \
--mapping Representatives.mapping \
--annotation Representatives.annotation \
--covariate Covariate.tsv \
--variant C1-Variant.tsv \
--output C1-MWAS.tsv
```

## 14. Metagenome-wide Association Analysis Plots

### Input

`C1-MWAS.tsv`

### Output

`C1-MWAS.qq.pdf` `C1-MWAS.manhattan.pdf`

```bash
MetaCAT plotMWAS \
--mwas C1-MWAS.tsv \
-o C1-MWAS
```

## 15. Replicaton Analysis In An Independent Cohort

Only SNPs that are significant in the discovery cohort are tested in the replication cohort.

### Input

`Representatives.mapping`

`Representatives.annotation`

`C2-Abundance.tsv`

`Phenotype.tsv`

`Covariate.tsv`

`C2-Variant.tsv`

`C2-MWAS.tsv`

### Output

`C2-MWAS.tsv`

```bash
MetaCAT mwas \
--abundance C2-Abundance.tsv \
--phenotype Phenotype.tsv \
--mapping Representatives.mapping \
--annotation Representatives.annotation \
--covariate Covariate.tsv \
--variant C2-Variant.tsv \
--output C2-MWAS.tsv \
--mwas C1-MWAS.tsv
```
