# MetaCAT

Metagenome Clustering and Association Tool.

All analyzed datasets are available at [Zenodo](https://doi.org/10.5281/zenodo.15202956).

https://doi.org/10.5281/zenodo.15202956

## CAMI Datasets

|File|Description|
|:-|:-|
|cami.binny-2.2.15patch17.mapping-files.tar.gz|Clustering mapping files generated by `Binny`.<br>FASTA-formatted cluster files can be generated using the `mapping2cluster` command of `MetaCAT`.|
|cami.comebin-1.0.4.mapping-files.tar.gz|Clustering mapping files generated by `COMEBin`.<br>FASTA-formatted cluster files can be generated using the `mapping2cluster` command of `MetaCAT`.|
|cami.concoct-1.0.0.mapping-files.tar.gz|Clustering mapping files generated by `CONCOCT`.<br>FASTA-formatted cluster files can be generated using the `mapping2cluster` command of `MetaCAT`.|
|cami.metabat2-2.12.1.mapping-files.tar.gz|Clustering mapping files generated by `MetaBAT2`.<br>FASTA-formatted cluster files can be generated using the `mapping2cluster` command of `MetaCAT`.|
|cami.metacat-1.0.3.mapping-files.tar.gz|Clustering mapping files generated by `MetaCAT`.<br>FASTA-formatted cluster files can be generated using the `mapping2cluster` command of `MetaCAT`.|
|cami.metadecoder-1.0.3.mapping-files.tar.gz|Clustering mapping files generated by `MetaDecoder`.<br>FASTA-formatted cluster files can be generated using the `mapping2cluster` command of `MetaCAT`.|
|cami.semibin2-1.5.1.mapping-files.tar.gz|Clustering mapping files generated by `SemiBin2`.<br>FASTA-formatted cluster files can be generated using the `mapping2cluster` command of `MetaCAT`.|
|cami.taxvamb-5.0.4.mapping-files.tar.gz|Clustering mapping files generated by `TaxVAMB`.<br>FASTA-formatted cluster files can be generated using the `mapping2cluster` command of `MetaCAT`.|
|cami.vamb-5.0.4.mapping-files.tar.gz|Clustering mapping files generated by `VAMB`.<br>FASTA-formatted cluster files can be generated using the `mapping2cluster` command of `MetaCAT`.|
|cami.metacat-1.0.3.processed-files.tar.gz|Seed and coverage files generated by the `seed` and `coverage` commands of `MetaCAT`.|

## IMG datasets

|File|Description|
|:-|:-|
|img.binny-2.2.15patch17.mapping-files.tar.gz|Clustering mapping files generated by `Binny`.<br>FASTA-formatted cluster files can be generated using the `mapping2cluster` command of `MetaCAT`.|
|img.comebin-1.0.4.mapping-files.tar.gz|Clustering mapping files generated by `COMEBin`.<br>FASTA-formatted cluster files can be generated using the `mapping2cluster` command of `MetaCAT`.|
|img.concoct-1.0.0.mapping-files.tar.gz|Clustering mapping files generated by `CONCOCT`.<br>FASTA-formatted cluster files can be generated using the `mapping2cluster` command of `MetaCAT`.|
|img.metabat2-2.12.1.mapping-files.tar.gz|Clustering mapping files generated by `MetaBAT2`.<br>FASTA-formatted cluster files can be generated using the `mapping2cluster` command of `MetaCAT`.|
|img.metacat-1.0.3.mapping-files.tar.gz|Clustering mapping files generated by `MetaCAT`.<br>FASTA-formatted cluster files can be generated using the `mapping2cluster` command of `MetaCAT`.|
|img.metadecoder-1.0.3.mapping-files.tar.gz|Clustering mapping files generated by `MetaDecoder`.<br>FASTA-formatted cluster files can be generated using the `mapping2cluster` command of `MetaCAT`.|
|img.semibin2-1.5.1.mapping-files.tar.gz|Clustering mapping files generated by `SemiBin2`.<br>FASTA-formatted cluster files can be generated using the `mapping2cluster` command of `MetaCAT`.|
|img.taxvamb-5.0.4.mapping-files.tar.gz|Clustering mapping files generated by `TaxVAMB`.<br>FASTA-formatted cluster files can be generated using the `mapping2cluster` command of `MetaCAT`.|
|img.vamb-5.0.4.mapping-files.tar.gz|Clustering mapping files generated by `VAMB`.<br>FASTA-formatted cluster files can be generated using the `mapping2cluster` command of `MetaCAT`.|
|img.metacat-1.0.3.processed-files.tar.gz|Seed and coverage files generated by the `seed` and `coverage` commands of `MetaCAT`.|
|img.checkm2-1.0.1.tsv.gz|Genome quality estimates generated by the `checkm2` command of `MetaCAT`.|
|img.gunc-1.0.6.tsv.gz|Genome quality estimates generated by the `GUNC`.|
|img.gtdbtk-2.3.2.tsv.gz|Taxonomic classifications generated by the `gtdbtk` command of `MetaCAT`.|

## CRC datasets

|File|Description|
|:-|:-|
|crc.assemblies.tar.xz.*|Metagenomic assemblies generated using `MegaHit`.<br>Reconstruct the the full archive with `cat crc.assemblies.tar.xz.* > crc.assemblies.tar.xz`.|
|crc.metacat-1.0.3.processed-files.tar.gz|Seed and coverage files generated by the `seed` and `coverage` commands of `MetaCAT`.|
|crc.metacat-1.0.3.mapping-files.tar.gz|Clustering mapping files generated by the `cluster` command of `MetaCAT`.|
|crc.checkm2-1.0.1.tsv.gz|Genome quality estimates generated by the `checkm2` command of `MetaCAT`.|
|crc.gtdbtk-2.3.2.tsv.gz|Taxonomic classifications generated by the `gtdbtk` command of `MetaCAT`.|
|crc.representative.assembly.gz|Representative genome file generated by the `representative` command of `MetaCAT`.|
|crc.representative.annotation.gz|Representative genome annotation file generated by the `representative` command of `MetaCAT`.|
|crc.representative.mapping.gz|Representative genome mapping file generated by the `representative` command of `MetaCAT`.|
|crc.715.abundance.gz|Abundacne table generated by the `abundance` command of `MetaCAT`.|
|crc.715.variant.gz|Variant file generated by the `variant` command of `MetaCAT`.|
|crc.715.mwas.gz|MWAS file generated by the `mwas` command of `MetaCAT`.|
|crc.166.abundance.gz|Abundacne table generated by the `abundance` command of `MetaCAT`.|
|crc.166.variant.gz|Variant file generated by the `variant` command of `MetaCAT`.|
|crc.166.mwas.gz|MWAS file generated by the `mwas` command of `MetaCAT`.|
